lib/stack/src/phantom_data.rs

```diff
@@ -0,0 +1,39 @@
+use std::marker::PhantomData;
+
+pub fn layer<T, M>() -> Layer<T, M>
+where
+    M: super::Stack<T>,
+{
+    Layer(PhantomData)
+}
+
+#[derive(Clone, Debug)]
+pub struct Layer<T, M>(PhantomData<fn() -> (T, M)>);
+
+#[derive(Clone, Debug)]
+pub struct Stack<T, M> {
+    inner: M,
+    _p: PhantomData<fn() -> (T, M)>,
```

Comment:
Nit: this should only need a `PhantomData(fn() -> T)`; `M` is bound by the `inner` field.

Also, TIOLI: adding a `where M: super::Stack<T>` bound to the definition of this struct would mean you don't need the `PhantomData` at all.

Reply:
> Also, TIOLI: adding a where M: super::Stack<T> bound to the definition of this struct would mean you don't need the PhantomData at all.

This is [not true](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=c9874cf1ef78b0b7ec4895ca5252a569).

Reply:
Oh, color me wrong, then! Sorry!